scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1493 @@
+            # verify n_samples is 2**n
+            if not (n_samples & (n_samples - 1) == 0):
+                warnings.warn("The balance properties of Sobol' points require"
+                              " n_samples to be a power of 2.")
+
+            if n_samples == 1:
+                sample = self._first_point
+            else:
+                _draw(n_samples - 1, self.num_generated, self.dim, self._sv,
+                      self._quasi, sample)
+                sample = np.concatenate([self._first_point,
+                                         sample])[:n_samples]
+        else:
+            _draw(n_samples, self.num_generated - 1, self.dim, self._sv,
+                  self._quasi, sample)
+
+        self.num_generated += n_samples
+        return sample
+
+    def random_base2(self, m=1):
+        """Draw point(s) from the Sobol' sequence.
+
+        This function draws :math:`n=2^m` points in the parameter space
+        ensuring the balance properties of the sequence.
+
+        Parameters
+        ----------
+        m : int
+            Logarithm in base 2 of the number of samples; i.e., n = 2^m.
+
+        Returns
+        -------
+        sample : array_like (n_samples, dim)
+            Sobol' sample.
+
+        """
+        n_samples = 2 ** m
+
+        total_n_samples = self.num_generated + n_samples
+        if not (total_n_samples & (total_n_samples - 1) == 0):
+            raise ValueError("The balance properties of Sobol' points require "
+                             "n to be a power of 2. {0} points have been "
+                             "previously generated, then: n={0}+2**{1}={2}. "
+                             "If you still want to do this, the function "
+                             "'Sobol.random()' can be used."
+                             .format(self.num_generated, m, total_n_samples))
+
+        return self.random(n_samples=n_samples)
+
+    def reset(self):
+        """Reset the engine to base state.
+
+        Returns
+        -------
+        engine: Sobol
+            Engine reset to its base state.
+
+        """
+        self._quasi = self._shift.copy()
+        self.num_generated = 0
+        return self
+
+    def fast_forward(self, n):
+        """Fast-forward the sequence by `n` positions.
+
+        Parameters
+        ----------
+        n: int
+            Number of points to skip in the sequence.
+
+        Returns
+        -------
+        engine: Sobol
+            The fast-forwarded engine.
+
+        """
+        if self.num_generated == 0:
+            _fast_forward(n - 1, self.num_generated, self.dim,
+                          self._sv, self._quasi)
+        else:
+            _fast_forward(n, self.num_generated - 1, self.dim,
+                          self._sv, self._quasi)
+        self.num_generated += n
+        return self
+
+
+def multinomial_qmc(n_samples, pvals, engine=None, seed=None):
+    """Draw low-discreancy quasi-random samples from multinomial distribution.
```

Comment:
Why is this a function whereas MultivariateNormal is a class? I would suggest consistency, if possible.